### ISSUE TITLE
Aastra + SRTP/RTP no audio after hold

### ISSUE THREAD

**Issue by GGGO** [sipwise/rtpengine#102]
Hi Richard,

we already talked about it, and I would like to create an issue for it.

Aastra ---SRTP---> rtpengine ---RTP---> rtpproxy---RTP--->asterisk

Note: rtpproxy will be removed soon, but for now I can't.

The problem is the following:
A calls B(Aastra), B(Aastra) answers A, 
B(Aastra) receives another call from C, B(Aastra) puts on hold A and answers C
B(Aastra) put on hold C and comes back to A
And now B(Aastra) can't hear A, but A can hear B(Aastra)

When I listen the call recording in asterisk, I can hear A and B(Aastra), so audio arrive without any problem.

It's difficult to reproduce the problem, I could make 10 or more tests without seing it.
It's happens sometimes for whatever reason.

But I've found this https://issues.asterisk.org/jira/browse/ASTERISK-20194, and it appears to be the same problem I have. Matt Jordan resolved it by applying 2 patches. See srtp_patches.diff

It seems Aastra change its crypto key, and I think rtpengine doesn't reset the crypto key if the endpoint hasn't change. So rtpengine will use on old crypto key, and Aastra will never success to decrypt the audio.

static void __fill_stream(struct packet_stream *ps, const struct endpoint *epp, unsigned int port_off) {
	struct endpoint ep;

	ep = *epp;
	ep.port += port_off;

	/* if the endpoint hasn't changed, we do nothing */
	if (PS_ISSET(ps, FILLED) && !memcmp(&ps->advertised_endpoint, &ep, sizeof(ep)))
		return;

	ps->endpoint = ep;
	ps->advertised_endpoint = ep;
	/* we reset crypto params whenever the endpoint changes */
	crypto_reset(&ps->crypto);
	dtls_shutdown(ps);
	PS_SET(ps, FILLED);
}

Maybe we need to keep the remote crypto key somewhere in the rtpengine session, and if the crypto has changed, we reset it ?

**Comment by GGGO** [comment]
I just commented this 2 lines to force crypto_reset() every times:

/*if (PS_ISSET(ps, FILLED) && !memcmp(&ps->advertised_endpoint, &ep, sizeof(ep)))
    return;*/

Will see if it's better.

**Comment by thomaswardiii** [comment]
The Asterisk bug appears to be the opposite of what you are experiencing.

In your bug: Alice calls Bob. Bob puts Alice on hold to answer Chris. Bob puts Chris on hold to talk to Alice. Bob can't hear Alice but Alice can hear Bob.

In the Asterisk bug:

Alice and Bob are in a call. After 10-15 minutes, Bob puts Alice on hold. After Bob resumes the call Alice cannot hear Bob but Bob can hear Alice.

**Comment by GGGO** [comment]
You are right.

I made a lot of tests since I commented the 2 lines, and the problem never came back but it's too early to conclude.

Let me test few days, and I will let you know.

**Comment by GGGO** [comment]
Bad news, the problem is still there...

Alice calls Bob and talks for 10min, Bob put Alice on hold, Bob unhold Alice, Bob can't hear Alice.

**Comment by thomaswardiii** [comment]
Does your issue exist after a period of time only? Or is it independent upon how long the call has been up?

Also does this always require a hard phone in the mix or can two SIP.js clients also result in the same behavior?

**Comment by marcantonio** [comment]
We are also seeing this issue with WebRTC clients and rtpengine.

**Comment by rfuchs** [comment]
As usual, the full rtpengine log would be helpful.

**Comment by GGGO** [comment]
In my case, I only use hard phone like Aastra which support SRTP.

And no, the problem could happen at any time.

Let me get the log.

**Comment by marcantonio** [comment]
Our scenario:

`[browser] <---> [rtpengine] <---> [asterisk] <---> [PSTN]`

If the client waits approximately 10-15 minutes, then puts the call on hold and resumes, the far side cannot be heard anymore.  I've done packet captures between rtpengine and Asterisk and I can confirm the audio can be heard in both directions.  Media is also being sent back to the client by rtpengine but I can't listen to it because it's encrypted (@rfuchs, this is why I was asking about decrypting the stream last week!).

Here's a log.  Line 63 is when the call is put on hold.

https://gist.github.com/marcantonio/e13a4a0ac845cf99709c

Thanks!

**Comment by GGGO** [comment]
Here the log:
https://github.com/GGGO/rtpengine/blob/master/rtpengine.log

Thanks!

**Comment by GGGO** [comment]
I would like to log the crypto key used by rtpengine when generating/sending the rtp packets to Aastra.

Is that possible ? I could compare with the crypto key asked by Aastra's phone.

I think rtpengine use the wrong crypto key.

**Comment by marcantonio** [comment]
If start rtpengine at log level 7 it will show you keys used for each stream.

**Comment by GGGO** [comment]
Rtpengine is already started at log level 7 but I just noticed that crypto_dump_keys() is called only for dtls.

I'm using SDES, it's why I don't have any keys logs.

Thanks for saying me that.

**Comment by rfuchs** [comment]
For SDES, the key can be found in the SDP dump. That's why it's not logged separately.

**Comment by rfuchs** [comment]
Can you check whether the SSRC changes or stays the same before and after the stream was put on hold?

**Comment by marcantonio** [comment]
The SSRC on the unencrypted farside (between asterisk and rtpengine)?

**Comment by rfuchs** [comment]
Either one of them :)

**Comment by GGGO** [comment]
The SSRC of Aastra changes after hold, and the SSRC of rtpengine stays the same.

It changes for every INVITE with SDP.

**Comment by marcantonio** [comment]
In the packet capture, I don't see the SSRC changing on either stream between asterisk and rtpengine.

**Comment by rfuchs** [comment]
Alright, so here's a _potential_ fix for SDES. Completely untested and just on a hunch, so fingers crossed.

```
diff --git a/daemon/crypto.h b/daemon/crypto.h
index 82e02dd..1311318 100644
--- a/daemon/crypto.h
+++ b/daemon/crypto.h
@@ -149,8 +149,26 @@ INLINE void crypto_params_copy(struct crypto_params *o, const struct crypto_para
                memcpy(o->mki, i->mki, i->mki_len);
        }
 }
+INLINE int crypto_params_cmp(const struct crypto_params *a, const struct crypto_params *b) {
+       if (a->crypto_suite != b->crypto_suite)
+               return 1;
+       if (!a->crypto_suite)
+               return 0;
+       if (memcmp(a->master_key, b->master_key, a->crypto_suite->master_key_len))
+               return 1;
+       if (memcmp(a->master_salt, b->master_salt, a->crypto_suite->master_salt_len))
+               return 1;
+       if (a->mki_len != b->mki_len)
+               return 1;
+       if (a->mki_len && memcmp(a->mki, b->mki, a->mki_len))
+               return 1;
+       return 0;
+}
 INLINE void crypto_init(struct crypto_context *c, const struct crypto_params *p) {
-       crypto_cleanup(c);
+       if (crypto_params_cmp(&c->params, p))
+               crypto_reset(c);
+       else
+               crypto_cleanup(c);
        crypto_params_copy(&c->params, p);
 }
 
```

**Comment by GGGO** [comment]
Ok patch applied, SRTP/RTP continues to work, hold/unhold works for my first calls.

Like you said, fingers crossed.

Now let's test !

**Comment by GGGO** [comment]
Problem still there, but this time I enabled the Aastra syslog and got this:
Apr 24 16:26:29 192.168.3.98 kernel: ept:srtpDecrypt: SMAPI_UnprotectRTP failed, errCode=-113
Apr 24 16:26:29 192.168.3.98 kernel: ept:srtpEgressStream: error, streamId = 0

That means Aastra failed to decrypt rtp packets, so that explains why there is no audio played.

I also tried with Zoiper 3.0 TLS+SRTP(SDES), and after a unhold, no voice but only noise.

**Comment by rfuchs** [comment]
I wonder if the Aastra phone expects the outgoing SRTP ROC to be reset to zero. Try changing crypto_init() in crypto.h to just:

```
INLINE void crypto_init(struct crypto_context *c, const struct crypto_params *p) {
	crypto_reset(c);
	crypto_params_copy(&c->params, p);
}
```

**Comment by GGGO** [comment]
Ok applied.

**Comment by marcantonio** [comment]
I tried the last change and got no audio in either direction after resuming.  The following was in the logs:

    Apr 24 14:16:26 cdeng-ccg-proxy rtpengine[15851]: [3c1ilb48mlm08vpct17t port 10042] SRTP output wanted, but no crypto suite was negotiated
    Apr 24 14:16:30 cdeng-ccg-proxy rtpengine[15851]: [3c1ilb48mlm08vpct17t port 10042] Confirmed peer address as 172.16.129.7:23950
    Apr 24 14:16:33 cdeng-ccg-proxy rtpengine[15851]: [3c1ilb48mlm08vpct17t port 10018] Confirmed peer address as 10.8.0.14:58513
    Apr 24 14:16:33 cdeng-ccg-proxy rtpengine[15851]: [3c1ilb48mlm08vpct17t port 10042] Kernelizing media stream
    Apr 24 14:16:33 cdeng-ccg-proxy rtpengine[15851]: [3c1ilb48mlm08vpct17t port 10042] No support for kernel packet forwarding available (encryption cipher or HMAC not supported by kernel module)
    Apr 24 14:16:41 cdeng-ccg-proxy rtpengine[15851]: [3c1ilb48mlm08vpct17t port 10042] SRTP output wanted, but no crypto suite was negotiated

**Comment by GGGO** [comment]
Problem still there :(

It easier to produce when I switch between lines. But again, it's completely random.

**Comment by GGGO** [comment]
I made a git clone 4 hours ago, and @rfuchs you commited a patch 3 hours ago, https://github.com/sipwise/rtpengine/commit/b0edf59a8f43b5c2eaf2062188b899537f37b93d

Do I need this patch too ? Because I don't have it.

**Comment by rfuchs** [comment]
It's unrelated, but feel free to merge it in anyway.

**Comment by GGGO** [comment]
I found this, very interesting. 

https://issues.asterisk.org/jira/browse/ASTERISK-16898

They speak about ROC and one way audio for some phone which don't support big endian for ROC. 

What do you think about it ?

**Comment by GGGO** [comment]
In hmac_sha1_rtp(), ROC is big endian (network order):
u_int32_t roc;
roc = htonl((index & 0xffffffff0000ULL) >> 16);
HMAC_Update(&hc, (unsigned char *) &roc, sizeof(roc));

Maybe we could try to convert it to little endian, but I don't know how to do that. My C skill is not that good !

**Comment by GGGO** [comment]
Good news ! It's not random at all ! I can reproduce the problem when I want.

First I tried with patches proposed here, not luck, so I cleaned up every thing to be sure and made a fresh git clone of master version and here my test.

The setup:
(Aastra1 and Aastra2) <---SRTP <---> rtpengine <---RTP <---> rtpproxy <---> RTP< ---> asterisk

Aastra1 and Aastra2 are registered to asterisk.

TEST1: FAILED
- Aastra1 calls Aastra2
- I'm watching the incoming rtp packets from rtpengine to my Aastra phones
- the Sequence number of packet from rtpengine to Aastra2 reset to zero
- at this point, audio is all fine
- Aastra1 puts on hold and unhold Aastra2
- then Aastra2 can't decrypt rtp packet, so Aastra2 can't hear Aastra1
- Aastra1 can hear Aastra2

TEST2: FAILED
- Aastra1 calls Aastra2
- the Sequence number of packet from rtpengine to Aastra1 reset to zero
- at this point, audio is all fine
- Aastra2 puts on hold and unhold Aastra1
- then Aastra1 can't decrypt rtp packet, so Aastra1 can't hear Aastra2
- Aastra2 can hear Aastra1

TEST3: OK
- Aastra1 calls Aastra2
- the Sequence number of packet from rtpengine to Aastra2 reset to zero
- at this point, audio is all fine
- Aastra2 puts on hold and unhold Aastra1
- audio is all fine on both

TEST4: OK
- Aastra1 calls Aastra2
- the Sequence number of packet from rtpengine to Aastra1 reset to zero
- at this point, audio is all fine
- Aastra1 puts on hold and unhold Aastra2
- audio is all fine on both

**Comment by GGGO** [comment]
So, here's the fix that works for me.

Aastra is strict about SSRC and ROC, when SSRC changes, ROC needs to reset to zero.
Asterisk will change the SSRC when doing hold/unhold thing.

Is this patch safe ? I don't really know, lock, memory leak,  c->ssrc need to be initialized ? Maybe there is a better way to code that.

And I don't know if I need to do the same in rtp_savp2avp...

Please let me know.

```
--- a/daemon/crypto.h  2015-04-25 23:51:22.296456005 -0400
+++ b/daemon/crypto.h     2015-04-25 23:52:35.512455306 -0400
@@ -80,6 +80,8 @@ struct crypto_context {
        void *session_key_ctx[2];

        int have_session_key:1;
+
+       u_int32_t ssrc;
 };


--- a/daemon/rtp.c     2015-04-25 23:51:22.298456005 -0400
+++ b/daemon/rtp.c        2015-04-25 23:53:34.375454745 -0400
@@ -189,6 +189,11 @@ int rtp_avp2savp(str *s, struct crypto_c
        if (check_session_keys(c))
                return -1;

+       if (c->ssrc != rtp->ssrc) {
+                c->last_index = 0;
+                c->ssrc = rtp->ssrc;
+        }
+
        index = packet_index(c, rtp);

        /* rfc 3711 section 3.1 */
```

**Comment by GGGO** [comment]
This case doesn't work yet:

- Aastra1 calls Aastra2
- The Seq number from rtpengine to Aastra1 reset to zero
- Aastra1 puts on hold/unhold Aastra2
     - Aastra1 has changed its SSRC, but Aastra2 not
- Aastra1 can't hear Aastra2

Is there any way I could reset ROC to zero as soon any of SSRC changes ?

Thank you.

**Comment by rfuchs** [comment]
Well, I did suspect that the ROC should be reset to zero, which is why I'm surprised that the patch from https://github.com/sipwise/rtpengine/issues/102#issuecomment-95992260 doesn't work, because it does exactly that.

As for standard compliance, I do agree that a change in SSRC must reset the ROC to zero for outgoing packets. But from your last comment, it appears that your setup also expects the ROC to reset to zero when the SSRC on the other side changes, which I don't think is correct behaviour.

**Comment by marcantonio** [comment]
I have a new bit of information.  @rfuchs, if you think I'm diverging from @GGGO's issue feel free to tell me to open a separate issue.

So, my issue is still as I explained in https://github.com/sipwise/rtpengine/issues/102#issuecomment-95689307.  The new bit is that if the browser client transfers the call to a third party (after resuming from hold results in no incoming audio), audio suddenly works again.  I don't know if this more firmly points at rtpengine or not.

Thanks.

**Comment by rfuchs** [comment]
Well, if you could confirm that this is either the same issue or a different issue by looking at a packet dump. The issue here is that the RTP sequence number wraps to zero, thus increasing the ROC, then the SSRC changes which should reset to ROC to zero but doesn't, which breaks decryption.

**Comment by marcantonio** [comment]
I can confirm that the sequence number wraps.  However, the SSRC does not change.  Also, we don't lose audio when it wraps, only later when we hold/unhold.

Is there any other info I can get?

**Comment by rfuchs** [comment]
Right, the SSRC would change when the audio comes back from being on hold and that's when audio would break.

**Comment by marcantonio** [comment]
So is the fact that the SSRC is not changing the issue?  If so, what is responsible for changing the SSRC?

**Comment by marcantonio** [comment]
I would guess that the reINVITE should change the SSRC in the SDP.  Is that correct?

**Comment by rfuchs** [comment]
No, the fact that the SSRC isn't changing means that it's a different issue. SSRC is part of the RTP header and determines the crypto context, which includes the ROC. Changing SSRC means ROC goes back to zero. If a client doesn't change the SSRC after on hold, then the ROC should remain the same.

**Comment by marcantonio** [comment]
I see.  So should the SSRC change when the sequence number wraps or after hold/resume?  Since the WebRTC client is ours we could make the latter happen.

**Comment by rfuchs** [comment]
If the SSRC changes, the ROC must be reset to zero. If it doesn't change, then ROC must remain unchanged.

Some clients change the SSRC when going on hold, others don't.

Seq wrapping around increases the ROC but doesn't affect anything else.

**Comment by marcantonio** [comment]
I was just starting think that the problem may be elsewhere when I came across this in the chrome debug logs:

    [40198:37635:0427/105354:VERBOSE2:srtpfilter.cc(576)] Failed to unprotect SRTP packet, err=7

right after resuming the call.  It seems to fall in line with the encryption breaking.

**Comment by GGGO** [comment]
@rfuchs, indeed the patch https://github.com/sipwise/rtpengine/issues/102#issuecomment-95992260 resolves this case:
- Aastra1 calls Aastra2
- The Seq number from rtpengine to Aastra1 reset to zero
- Aastra1 puts on hold/unhold Aastra2
- audio in both direction ok

but it breaks the case where:
- Aastra1 calls Aastra2
- The Seq number from rtpengine to Aastra2 reset to zero
- Aastra2 puts on hold/unhold Aastra1
- Aastra2 can't hear Aastra1

Note:
Aastra always changes its SSRC after a hold/unhold
Asterisk doesn't change its SSRC for the one who asked the hold/unhold
But asterisk will change its SSRC for the remote side connected with the one who asked the hold

**Comment by rfuchs** [comment]
OK, yes that makes sense.

**Comment by mclazarus** [comment]
Hello @rfuchs I'm working with @marcantonio investigating the problem on our end.  What I noticed today on reproducing is that when I put the call on hold the SSRC and Seq change.  And when I go off hold the SSRC goes back and the Seq becomes a lower value than it was after going on hold.  

I need to run a bunch 

Details:

RTP coming from asterisk right before hold is SSRC=0xF6E4640 Seq=26747

Right after going on hold SSRC=0xB8BC8D4 Seq=37169

Right before resuming (going off hold) SSRC=0xB8BC8D4 Seq=37828

Right after resuming (going off hold) SSRC=0xF6E4640 Seq=27409

I'm not sure at this point if this data is helpful.  But I'm going to do some more investigation on my end and see if I can find more to help me understand the failure here.

Thanks!

**Comment by rfuchs** [comment]
Can you tell what happens to the ROC on your end in these cases? Right now, rtpengine would reset the ROC to zero twice (each time the SSRC changes). Technically though, the crypto context at the end would be the same as at the beginning and so the ROC should retain its value between those two, however that's not supported in rtpengine right now.

**Comment by marcantonio** [comment]
I'm printing out some data each time `rtp_avp2savp()` runs.  I assume this is for every RTP packet.  I noticed the following:

- `c->last_index` (the ROC counter?) increments for every packet.
- `rtp->seq_num` seems to increment very quickly.  I expected this to correlate to the sequence number in the RTP packets, but it doesn't.
- `c->last_index` freezes at some point in the call.  It is a this point that I lose audio.  I'm not sure if this is another issue or not.

Those are just my observations as we dig in further.  The latter is the only one that really seems to be a problem at this point.  The others are likely my ignorance of RTP.

**Comment by GGGO** [comment]
@marcantonio could you show me your ilog debug, I will try to make the same.

**Comment by rfuchs** [comment]
c->last_index is an implicit ROC. It's the last RTP sequence number with the current ROC taken into account.

rtp->seq_num would be in network byte order. You'll have to swap it (ntohs) if you want to print it.

**Comment by marcantonio** [comment]
AH!  I forgot!

**Comment by marcantonio** [comment]
So I came across the last_index freezing again.  It did not correspond to a loss of audio:

```
c->ssrc:       413E29F1 c->last_index: 40140
rtp->ssrc:     413E29F1 rtp->seq_num:  40141
c->ssrc:       413E29F1 c->last_index: 40141
rtp->ssrc:     413E29F1 rtp->seq_num:  40142
c->ssrc:       413E29F1 c->last_index: 40142
rtp->ssrc:     413E29F1 rtp->seq_num:  40143
c->ssrc:       413E29F1 c->last_index: 40143
rtp->ssrc:     413E29F1 rtp->seq_num:  40144
c->ssrc:       413E29F1 c->last_index: 40144
rtp->ssrc:     413E29F1 rtp->seq_num:  40145
c->ssrc:       413E29F1 c->last_index: 7349
rtp->ssrc:     413E29F1 rtp->seq_num:  40146
c->ssrc:       413E29F1 c->last_index: 7349
rtp->ssrc:     413E29F1 rtp->seq_num:  40147
c->ssrc:       413E29F1 c->last_index: 7349
rtp->ssrc:     413E29F1 rtp->seq_num:  40148
c->ssrc:       413E29F1 c->last_index: 7349
rtp->ssrc:     413E29F1 rtp->seq_num:  40149
c->ssrc:       413E29F1 c->last_index: 7349
rtp->ssrc:     413E29F1 rtp->seq_num:  40150
c->ssrc:       413E29F1 c->last_index: 7349
rtp->ssrc:     413E29F1 rtp->seq_num:  40151
c->ssrc:       413E29F1 c->last_index: 7349
rtp->ssrc:     413E29F1 rtp->seq_num:  40152
c->ssrc:       413E29F1 c->last_index: 7349
rtp->ssrc:     413E29F1 rtp->seq_num:  40153
```

**Comment by rfuchs** [comment]
Can you share the code you use to print these?

**Comment by marcantonio** [comment]
https://gist.github.com/marcantonio/1b880bc9653029e0bfb9#file-gistfile1-txt-L11-L15

Very simple and I'm running rtpengine from the foreground.

**Comment by rfuchs** [comment]
I can't reproduce this. What platform is this on?

**Comment by marcantonio** [comment]
Linux proxy 3.13.0-43-generic #72-Ubuntu SMP Mon Dec 8 19:35:06 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux

It's an Ubuntu 12.04 VM.  I'm not dwelling on that issue as I don't think it relates to the problem at hand.  Everything and @mclazarus and I are looking at suggests that something (most likely the index) is not right when it's passed to `crypto_encrypt_rtp()` after a rollover.  We're looking through RFC3711 now.

It could very well be Chrome screwing up the ROC too.

**Comment by marcantonio** [comment]
I'm going to open another issue because one is not ours.  In fact, the patch that came of this (https://github.com/sipwise/rtpengine/commit/39a305f4badd3bc43d2d5239e0112006ce2540d2), breaks hold/resume right after any rollover.  Zeroing out the ROC doesn't seem like the right thing to do based on the RFC.

**Comment by rfuchs** [comment]
A zero ROC is correct for any new crypto context, and a new SSRC leads to a new crypto context. I agree that if the SSRC switches back and forth between two or more values, then the ROC will end up wrong because, as I said before, that's not supported right now.

**Comment by cahlbin** [comment]
@marcantonio @mclazarus

Did you find a resolution for this issue? (I saw you closed https://github.com/sipwise/rtpengine/issues/108 as a non-issue right (iptables not configured properly?))

I have a similar roll-over issue. I observe no change in SSRC whatsoever, and I don't make use of any hold-feature. Still the auth error kicks in when the packet sequence counter crosses over from 65536 to 0, and the ROC should increment. The setup is WebRTC Client <--SRTP --> RTPEngine <-- RTP --> Asterisk.

**Comment by rfuchs** [comment]
Can you tell which side is at fault?

**Comment by cahlbin** [comment]
It's the WebRTC client that rejects the packets sent from rtpengine. (The client is running WebRTC native (iOS), with libsrtp 1.5.0)

**Comment by rfuchs** [comment]
That's probably a bug in the WebRTC client then. But if you make a pcap of such a rollover, then we can confirm that rtpengine is correctly increasing the ROC when sending packets out. With the SRTP keys dumped in the log (log level 7), the encryption can easily be verified.

**Comment by mclazarus** [comment]
Hi @cahlbin 

We are continuing to investigate.  Our diagram is basically the same as yours:

WebRTC Client <-- SRTP --> rtpengine <-- RTP --> asterisk <-- RTP --> VoIP Phone

The root of our issue appears to be that Asterisk will change SSRC in three cases.

1. the remote end e.g. VoIP Phone in the following diagram puts the call on hold and we start hearing hold music.  
2. the remote end resumes the call restoring the SSRC from prior to putting us on hold.
3. The webrtc end initiates hold sends a REINVITE via kamailio to asterisk.
4. the webrtc end resumes the call sending a REINVITE via kamailio to asterisk.

Our webrtc clients have been both Firefox and Chrome.  We see Chrome is using libsrtp internally (not sure which version).  Not sure what Firefox is using, but I'd guess they rolled their own (I haven't investigated)

Situation 1 is a problem because rtpengine doesn't unkernelize the call because it has no reason to, the SSRC and sequence number change but the rtpengine code just treats this as dropped packets on the same stream.  This can cause in correct rolling forward of the roc, or even taking the roc negative.  Which libsrtp doesn't agree with, it seems to be tracking ssrc changes on a stream by preserving/restoring the roc or zeroing it.

Situation 2 is a problem because the old SSRC and sequence numbers are resumed again the packets remain kernelized and the the old packet_index is not restored on the rtpengine side.  Again this causes the packet_index() algorithm to incorrectly manipulate the roc because of changes in the sequence number in what rtpengine assumes is the same stream.

Situation 3 causes the call to fall back to userspace and be handled.  The recent patch referenced in this issue actually 0s out the roc in that case, which works okay for the silence Asterisk sends me when I initiate hold, but is not good when I restore the call as in Situation 4, because if I had accumulated the roc on that first call it will now be zero and in disagreement with libsrtp.

Luckily chrome/libsrtp doesn't play audio when the hash fails, where firefox plays you a nice symphony of noise.

Basically we've been working full time on our end to see if we can introduce basic ssrc tracking into rtpengine and preserve the roc and restore it when the ssrc changes.  We think we have a good workable solution in userspace but the kernel side is a bit more worrying.

Depending on your client, if you're using chrome you can pass:

`--enable-logging --v=4 --vmodule=*libjingle/source/talk/*=4` to chrome to have it log the encryption errors and learn a little more about what is going on.  I'd guess 

Also @rfuchs I have a couple questions if you have time.

1. When a call is in kernelspace should the daemon end see the packets as well?  Some of our tests have shown that and I found it confusing because it looks like the rtpengine46() returns NF_DROP.  I assumed that would drop the packet before it was passed on to userspace but our tests aren't showing that.

2. I see the rtpengine packet_index() algorithm differs from the pseudocode in the RFC, but I couldn't find reasoning in commit messages for why it's been modified.  Can you remember or shed any light on to why it is different?  I see The algorithm current in the code has some dead spots in the difference between seq and the seq part of packet_index where packet_index isn't updated at all but I can't figure out why that would be useful.

Thanks.

**Comment by rfuchs** [comment]
1. Normally the userspace daemon doesn't get to see the packets, the kernel module forwards them and will then drop the original packet. However, the userspace daemon acts as a fallback if the kernel module can't handle the packet for some reason, for example if SRTP authentication fails. The daemon is then supposed to handle the packet in some way, which may include logging an error message.

2. The code is different because it doesn't track the ROC as a separate value (as the RFC suggests) but it rather tracks the sequence number as a 64-bit number, which implicitly includes the ROC in the high-order bits. The "dead spots" you mention are probably cases where the received sequence number is actually below the last_index, indicating packets received out of order.

**Comment by mclazarus** [comment]
As we are testing working out a way to handle changes in SSRC on a single stream preserving the packet_index do you have any guidance as to what you would find acceptable for inclusion into the codebase?

**Comment by rfuchs** [comment]
Something like a hash table would probably work. For the kernel module, it might be easiest to just lock the last known SSRC in place and use the userspace fallback when a SSRC mismatch is seen. The daemon would then have to update the kernel info when that happens. But I'm open to suggestions and other ideas.

**Comment by mclazarus** [comment]
Thanks.  I found the bug we introduced dropping all encryption packets back to userspace.  And the ideas you said are similar to what we were testing with.  We've been building a linked list of structs that we key and search by ssrc, and we hold the packet_index in there.  

In order to `unkernelize` I assume we'll have to go to the error state in `rtpengine46` when the ssrc mismatch occurs and then from userspace detect that we should call MM_DEL so all packets will be in userspace until we clean up and it calls `kernelize()` again.  Does this sound right?

**Comment by rfuchs** [comment]
Yeah, sounds about right. I thought it should be possible to directly update the kernel info without going through the delete/re-add process, but I can't tell right now if that's actually feasible.

**Comment by mclazarus** [comment]
We've just successfully completed some testing handling SSRC changes so should have a pull request for review tomorrow.

**Comment by GGGO** [comment]
@mclazarus I tried your patch and it didn't resolve my problem.

I think @rfuchs is right, Aastra phone seems to reset all ROC in both direction if one SSRC changes.

Aastra alias Mitel works on it, and tries to determine if the problem is the server side or the phone.

I'm waiting for good news.

**Comment by GGGO** [comment]
I got a follow up from Aastra:

"As the phone uses a new SSRC for the stream from the phone to the server ("outgoing" stream) after the hold/retrieve, the expected behaviour is it would reset the ROC for the "outgoing" stream. Since the "outgoing" audio is fine, there is no issue here.

However, as the "incoming" stream SSRC is unchanged, the expected behaviour is it would not reset the ROC for the "incoming" stream. Since there "incoming" audio stops, this behaviour is what needs to be investigated to determine what end (phone or server) is behaving incorrectly.

The phone *is* designed to behave in the manner described, so at first glance I would suspect the server is incorrectly resetting the ROC on it's side for the stream from the server to the phone. However, to investigate further, I need to make test calls and review my captured traces and syslog."

So I need to trace the ROC to be sure that rtpengine is not resetting it to 0. Is it ok to trace last_index ?

**Comment by rfuchs** [comment]
Sure, or you can do it through a pcap. If you have a pcap and the SRTP keys, you can use the auth tag to verify that the encryption is correct (which includes verification of the ROC)

**Comment by mclazarus** [comment]
@GGGO when we were troubleshooting this we would drop `printk()`s into the rtpengine46() or packet_index() functions taking care to figure out which side was encrypt and decrypt to see the packet_index.

Although be aware when we were testing with asterisk we'd see that the SSRC would change when going on hold from the phone (to a stream with just silence in it) and then change back to the old stream when we'd take the call off hold.  It was there we saw rtpengine zeroing the ROC, and what our patch prevents from happening.  Possibly the AASTRA is doing the same thing that rtpengine was doing.  We saw that both chrome (libsrtp) and firefox did not zero the ROC in that case.

**Comment by GGGO** [comment]
Oh ! And Aastra uses libsrtp too.

Thank you to say me that !

**Comment by GGGO** [comment]
I made a similar test with a Yealink-T28P, after hold/unhold the phone can't decrypt RTP packets.

So we got the same behavior with Aastra, Yealink and Zoiper... I know that it's not RFC compliant but maybe rtpengine need implement a special option for these phones.

**Comment by GGGO** [comment]
Does the last commit help with this ?

**Comment by rfuchs** [comment]
Probably not, that's why I reopened it.

**Comment by GGGO** [comment]
With https://github.com/sipwise/rtpengine/commit/f398d837df223d9a984c656467c1c3fa25e70d94 something has changed !

Now here what is happenning, (in one call):
- Aastra1 calls Aastra2
- Aastra2 answers Aastra1
- SEQ from rtpengine to Aastra2 resets to 0
- Aastra2 puts on hold/unhold Aastra1
- two way audio !! Good news !
- Aastra1 puts on hold/unhold Aastra2
- One way audio, Aastra1 can hear Aastra2 BUT Aastra2 can't hear Aastra1. :(
Note: to have audio back on Aastra2, I have to complety take line 2 and make a new call. After hangup the line 2, I come back to line1 and audio is back. But multiple hold/unhold never help me to get the audio back.


Another test:
- Aastra1 calls Aastra2
- Aastra2 answers Aastra1
- SEQ from rtpengine to Aastra2 resets to 0
- Aastra1 puts on hold/unhold Aastra2
- two way audio.

I could test with Yealink too, to be sure that is not specific to Aastra.

So, what do you think about that ?

Thanks.

**Comment by rfuchs** [comment]
From what you've reported, I suppose what's missing to make it work for you is to reset the ROC also in the opposite direction when the SSRC changes.

**Comment by GGGO** [comment]
Mitel has confirmed the issue.

I'm waiting for the new firmware.

**Comment by GGGO** [comment]
Hi ! It's me again on this issue...

Mitel hasn't released any patches for this bug, so I got an idea !

I installed Asterisk-1.8.x in a virtualbox and connected 2 Aastra on it. And guess what ! Hold/unhold works perfectly ! But I can't use asterisk directly and need rtpengine in my setup.

So, I tried to find how Asterisk solves this and I got some information in changelog:

https://reviewboard.asterisk.org/r/1741/diff/#index_header
https://reviewboard.asterisk.org/r/1885/diff/#index_header
https://reviewboard.asterisk.org/r/2099/diff/#index_header

Does someone understand how they did it ?

I hope rtpengine could handle this too, it's not specific to Aastra, but Yealink too.

Thank you !

**Comment by GGGO** [comment]
Mitel has released a new firmware and the issue DEF40208 was fixed:

"In certain environments, the phone incorrectly reset the SRTP Rollover Counter (RoC) to zero when attempting to retrieve a held call on one line while on an active call on another line.
This issue has been corrected."

I didn't try the new release yet, but this issue is definitely not related to rtpengine, so I'm closing this.
